add scores APIs, db schema and queries

ravikp7 opened the pull request.

ravikp7 commented:
- Add APIs to get individual user scores, scores in a range and save score
- Add DB schema and queries for operations
- Tests for DB schema and route controllers

fixes #12

ravikp7 commented:
`CI is not added yet, so attaching screenshot for passing tests` Will be adding CI next
![screenshot from 2019-02-19 05-17-34](https://user-images.githubusercontent.com/13567359/52981446-dab0a000-3405-11e9-9716-b1d638b4b52f.png)

ravikp7 commented:
- DB query tests are not implemented yet as it's dependent on DB setup and config which are to be implemented by @adityasmksaxena 
- Will be adding after that

ravikp7 commented:
Tested the DB with own server/index.js file with changes which I've not committed here

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  HTTP_SUCCESS_OK,
+  HTTP_INTERNAL_SERVER_ERROR,
+  HTTP_CREATED,
+} = require('../../utils/contants');
```
why not import statements?

ravikp7 replied:
Since, the current Node LTS doesn't support import/export and we didn't setup Babel, so went with it. I understand the advantages of using import/export and once support is there, we can just drop the Babel transpilation. Will add webpack and babel

nimish-gupta replied:
ok good call for that.
Please add babel for latest ES features, why you need webpack?

ravikp7 replied:
Yes, eslint is setup with the airbnb config

ravikp7 replied:
> ok good call for that.
> Please add babel for latest ES features, why you need webpack?

I thought webpack is required for babel transpilation with nodemon. Checked it now, it isn't the case. Thank you :) :100:

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  getScores,
+  getUserScores,
+  createScore,
+} = require('../../controllers/score.controller');
```
why not import statements?

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  queryScores,
+  queryUserScore,
+  insertScore,
+} = require('../../utils/db_score_query');
```
why not import statements?

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+const error = new Error('Query Failed');
+
+// Mocked DB function to get scores in range
+queryScores.mockImplementation(range => new Promise(
```
Why not `async/await`?

ravikp7 replied:
There, in that function, I had to `return a promise which resolves or rejects based on a conditional`. I'm not sure how using `async/await` can make it better. Can you please elaborate more?

nimish-gupta replied:
```suggestion
queryScores.mockImplementation(async (range) => {
    if (range) {
        return range
    }
    throw new Error('Range not defined')
})
```

ravikp7 replied:
I wasn't aware of this `Implicit Promise` behaviour of async functions. Thanks for this :+1:

nimish-gupta replied:
Check [this](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function)

ravikp7 replied:
Yes, had already checked it, thanks :+1:

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  },
+));
+
+queryUserScore.mockImplementation(userId => new Promise(
```
Why not `async/await`?

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  },
+));
+
+insertScore.mockImplementation((userId, value) => new Promise(
```
Why not `async/await`?

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+    jsonData: '',
+    status(statusCode) {
+      this.responseStatus = statusCode;
+      function json(data) {
```
arrow functions?

ravikp7 replied:
I'd to bind that function, so my first approach was to use normal function and bind that. I didn't realize that arrow functions are there for this purpose. Will replace it with arrow function now. Thank you :)

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  };
+
+  describe('getScores GET Controller', () => {
+    it('should call db function with range', async () => {
```
You can use `beforeEach` hook to call `getScores` for every success request and keep that in describe block. A separate describe block for every invalid request and a beforeEach hook.

ravikp7 replied:
I thought of testing all the functionality of single controller in a single describe block. Though haven't really used `beforeEach` before, will definitely try to use it next time. Thanks :)

nimish-gupta commented on server/src/utils/db_score_query.js:
```diff
@@ -0,0 +1,26 @@
+  .exec();
+  // exec() is used to get a Promise
+  // Mongoose queries return a thenable object, but it isn't a Promise
+  // https://mongoosejs.com/docs/promises.html#queries-are-not-promises
```
👍

nimish-gupta commented on server/src/controllers/score.controller.js:
```diff
@@ -0,0 +1,52 @@
+
+const getScores = (req, res) => {
+  const { range } = req.params;
+  return queryScores(parseInt(range, 10))
```
Why not `async/await`?

nimish-gupta commented on server/src/controllers/score.controller.js:
```diff
@@ -0,0 +1,52 @@
+
+const getUserScores = (req, res) => {
+  const { userId } = req.params;
+  return queryUserScore(userId)
```
Why not `async/await`?

nimish-gupta commented on server/src/controllers/score.controller.js:
```diff
@@ -0,0 +1,52 @@
+
+const createScore = (req, res) => {
+  const { userId, value } = req.body;
+  return insertScore(userId, value)
```
Why not `async/await`?

nimish-gupta commented on server/src/controllers/score.controller.js:
```diff
@@ -0,0 +1,52 @@
+const scoreDbQuery = require('../utils/db_score_query');
```
why not import statements?

nimish-gupta commented on server/src/controllers/score.controller.js:
```diff
@@ -0,0 +1,52 @@
+  HTTP_SUCCESS_OK,
+  HTTP_INTERNAL_SERVER_ERROR,
+  HTTP_CREATED,
+} = require('../utils/contants');
```
why not import statements?

nimish-gupta commented on server/src/models/score.model.js:
```diff
@@ -0,0 +1,18 @@
+const mongoose = require('mongoose');
```
why not import statements?

nimish-gupta commented on server/src/routes/score.route.js:
```diff
@@ -0,0 +1,16 @@
+const express = require('express');
+const scoreController = require('../controllers/score.controller');
```
why not import statements?

nimish-gupta commented on server/src/routes/score.route.js:
```diff
@@ -0,0 +1,16 @@
+const express = require('express');
```
why not import statements?

nimish-gupta commented on server/src/__tests__/models/score.model.test.js:
```diff
@@ -0,0 +1,17 @@
+const Score = require('../../models/score.model');
+
+describe('leaderboard schema validation', () => {
```
Its good that you are checking the schema validation, love the way that how you tested, but it seems that these are overkill over here because schema is provided by mongoose library and you are using functions provided by mongoose itself, and you are not creating any of your own function so basically you are checking mongoose functionality which is already tested by the library mongoose itself, so basically you don't need these tests here.

ravikp7 replied:
The intention was to ensure that in the schema definition, the field is set to required. As, it's very easy to miss that required in the field. Will it be still an overkill? Let me know your thoughts :)

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  });
+
+  describe('getUserScores GET Controller', () => {
+    it('should call db function with userID', async () => {
```
You can use `beforeEach` hook to call `getScores` for every success request and keep that in describe block. A separate describe block for every invalid request and a beforeEach hook.

nimish-gupta commented on server/src/__tests__/controllers/score.controller.test.js:
```diff
@@ -0,0 +1,168 @@
+  });
+
+  describe('createScore POST Controller', () => {
+    it('should call db function with userId and value', async () => {
```
You can use `beforeEach` hook to call `getScores` for every success request and keep that in describe block. A separate describe block for every invalid request and a beforeEach hook.

nimish-gupta reviewed (changes requested):
@ravikp7 have you guys setup the eslint on the server side?

ravikp7 committed:
chore: add babel for server

It enables es6 import/export usage in server code

fixes #13

ravikp7 committed:
feat: add score api and db schema

    - Add APIs to get individual user scores, scores in a range and save score
    - Add DB schema and queries for operations
    - Tests for DB schema and route controllers

fixes #12

ravikp7 commented:
`Tests Screenshot`
![screenshot from 2019-02-20 00-04-46](https://user-images.githubusercontent.com/13567359/53039053-3b8eb580-34a4-11e9-87e1-0719b4336117.png)

ravikp7 commented:
Fixed in this pull request after review:

`First Commit`
- Added babel for server

`Second Commit`
- Used es6 import/export instead of require
- Replaced the bound function with arrow one
- Used `beforeEach` for each controller using separate describe blocks for `SUCCESS` and `FAILURE`

nimish-gupta commented:
@ravikp7 you can replace `.then` with `async/await` syntax in your code.

ravikp7 commented:
> @ravikp7 you can replace `.then` with `async/await` syntax in your code.

@nimish-gupta done :+1:

nimish-gupta reviewed (approved):
LGTM 3>
